core/src/main/scala/org/apache/spark/util/Utils.scala

```diff
@@ -810,7 +810,7 @@ private[spark] object Utils extends Logging {
    * Get the local machine's URI.
    */
   def localHostNameForURI(): String = {
-    customHostname.getOrElse(InetAddresses.toUriString(localIpAddress))
+    sys.env.get("SPARK_PUBLIC_DNS").getOrElse( customHostname.getOrElse(InetAddresses.toUriString(localIpAddress)) )
```

Comment:
Probably worth revisiting this part of the code. I see I should have just relied on `SPARK_LOCAL_HOSTNAME`, however, at first thought, I had trouble getting it to work (it was reset by some other code, without me explicitly calling it, to another value).
I admit - I do not understand the distinction between `SPARK_PUBLIC_DNS` and `SPARK_LOCAL_HOSTNAME`. `SPARK_PUBLIC_DNS` is used by the driver only and the other one is used for all other services. Why the difference?

It may be just my assumption but if I'm setting `SPARK_PUBLIC_DNS`, I'd like to have all services bound to `SPARK_PUBLIC_DNS`.